### PR TITLE
fix(elements): update the view of an `OnPush` component when inputs change

### LINKED ISSUE #38948: Angular/elements OnPush input changes does not trigger change detection
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/elements

### Is this a regression?

 dunno

### Description

- create custom elements via @angular/elements from component with input and `changeDetection: ChangeDetectionStrategy.OnPush`
- set it's initial input value from the application or change it
- component's CD is not triggered, template is not updated

## 🔬 Minimal Reproduction

Clone https://github.com/artaommahe/Angular-elements-input-update. There are two applications: `widgets` with custom element and `main-app` with custom element usage in other angular app.

Two ways to face the issue - from `main-app` or from plain js custom element usage.

##### Plain js

platform's code here https://github.com/artaommahe/Angular-elements-input-update/blob/master/projects/widgets/src/index.html#L11

- run `widgets` dev server `yarn start widgets`
- open `http://localhost:4201/`
- click `inc` button multiple times
- see no changes in widget's template
- click `trigger widget CD` button (runs markForCheck inside widget component)
- see changes
- uncomment this https://github.com/artaommahe/Angular-elements-input-update/blob/master/projects/widgets/src/app/some-widget/components/some-widget/some-widget.component.ts#L17
- update page and click `inc` button
- see changes

##### Other angular app

platform's code here https://github.com/artaommahe/Angular-elements-input-update/blob/master/projects/main-app/src/app/app.component.ts#L6

- run `widgets` dev server `yarn start widgets --extra-webpack-config projects/widgets/webpack.config.js` (we cant run two angular-cli apps in one browser tab webpack config changes https://github.com/angular/angular-cli/issues/8902)
- run `main-app` dev server `yarn start main-app`
- open `http://localhost:4200/`
- click `inc` button multiple times
- see no changes in widget's template
- click `trigger widget CD` button (runs markForCheck inside widget component)
- see changes
- uncomment this https://github.com/artaommahe/Angular-elements-input-update/blob/master/projects/widgets/src/app/some-widget/components/some-widget/some-widget.component.ts#L17
- update page and click `inc` button
- see changes

or you can remove `ChangeDetectionStrategy.OnPush` in `some-widget` component and everything will work too

## 🌍  Your Environment

**Angular Version:**
<pre><code>
PS D:\projects\issues\angular-elements-input-update> yarn ng version                                                                  
yarn run v1.16.0
$ ng version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 10.1.2
Node: 12.18.1
OS: win32 x64

Angular: 10.1.2
... animations, cli, common, compiler, compiler-cli, core
... elements, forms, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1001.2
@angular-devkit/build-angular   0.1001.2
@angular-devkit/core            10.1.2
@angular-devkit/schematics      10.1.2
@schematics/angular             10.1.2
@schematics/update              0.1001.2
rxjs                            6.6.3
typescript                      4.0.3
</code></pre>

**Anything else relevant?**

### PR DESCRIPTION
**Pull request by gkalpak**
As with regular Angular components, Angular elements are expected to have their views update when inputs change.

Previously, Angular Elements views where not update if the underlying component used the `OnPush` change detection strategy.

This commit fixes this by calling `markForCheck()` on the component view's `ChangeDetectorRef`.

NOTE:
This is similar to how `@angular/upgrade` does it: https://github.com/angular/angular/blob/3236ae0ee118d0734c90fa9f3767435396213470/packages/upgrade/src/common/src/downgrade_component_adapter.ts#L146.

Fixes #38948.

##
This PR also contains some commit the fix/improve the `elements` tests (both unit and e2e tests). See individual commits for more details.

### CODE REVIEW

File: goldens/size-tracking/aio-payloads.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 3037,
-        "main-es2015": 448085,
+        "main-es2015": 448756,
```
Comment (gkalpak) [inline goldens/size-tracking/aio-payloads.json:15]: The actual increase here is 277B (from [448479][1] to [448756][2]).

[1]: https://app.circleci.com/pipelines/github/angular/angular/23665/workflows/48f3734b-5ed8-43d8-9b50-6c3bf0d21186/jobs/842748
[2]: https://app.circleci.com/pipelines/github/angular/angular/23709/workflows/82d2e0ea-a037-4ea3-a81d-edb625870cf4/jobs/843518

File: goldens/size-tracking/aio-payloads.json [hunk 2]
```diff
@@ -21,7 +21,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 3157,
-        "main-es2015": 431750,
+        "main-es2015": 432151,
```
Comment (gkalpak) [inline goldens/size-tracking/aio-payloads.json:24]: The actual increase here is 277B (from [431874][1] to [432151][2]).

[1]: https://app.circleci.com/pipelines/github/angular/angular/23665/workflows/48f3734b-5ed8-43d8-9b50-6c3bf0d21186/jobs/842746
[2]: https://app.circleci.com/pipelines/github/angular/angular/23709/workflows/82d2e0ea-a037-4ea3-a81d-edb625870cf4/jobs/843515

**Review by josephperrott — APPROVED** [review]
LGTM

**Review by petebacondarwin — COMMENTED** [review]
typo in the last commit:

`where not update` -> `were not updated`

**Review by petebacondarwin — APPROVED** [review]
LGTM 

Reviewed-for: size-tracking

**Review by IgorMinar — APPROVED** [review]
LGTM 

Reviewed-for: size-tracking

**Review by atscott — APPROVED** [review]

File: packages/elements/test/component-factory-strategy_spec.ts [hunk 3]
```diff
@@ -261,6 +264,55 @@ describe('ComponentFactoryNgElementStrategy', () => {
...
+
+    it('should mark the view for check once for multiple input changes', fakeAsync(() => {
+         strategy.setInputValue('fooFoo', 'fooFoo-1');
+         strategy.setInputValue('barBar', 'barBar-1');
+         tick(16);  // scheduler waits 16ms if RAF is unavailable
+
+         expect(viewChangeDetectorRef.markForCheck).toHaveBeenCalledTimes(1);
+       }));
+
+    it('should mark the view for check twice for changes in different rounds with previous values',
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/elements/test/component-factory-strategy_spec.ts:317]: Good questions. I've added tests for the first case (setting to the same value in a new round).

For the second case (setting to a new value and then back to the original value), I don't think we'll gain much by adding tests. There is no logic to detect that, so `detectChanges()` (including `markForCheck()`) will run (although redundant 🤷‍♂️).

**Comment by gkalpak** [comment]
NOTE: #39463 is trying to address the issue more generically.

File: goldens/size-tracking/aio-payloads.json [hunk 4]
```diff
@@ -12,7 +12,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 3037,
-        "main-es2015": 448615,
+        "main-es2015": 448922,
```
Comment (gkalpak) [inline goldens/size-tracking/aio-payloads.json:15]: The actual increase here is 277B (from [448645][1] to [448922][2]).

[1]: https://app.circleci.com/pipelines/github/angular/angular/24192/workflows/1f86d952-a813-4e00-8aba-11c2420c18dc/jobs/850932
[2]: https://app.circleci.com/pipelines/github/angular/angular/24189/workflows/03d37062-e9d7-401e-a563-7a386db54994/jobs/850877

File: goldens/size-tracking/aio-payloads.json [hunk 5]
```diff
@@ -21,7 +21,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 3157,
-        "main-es2015": 431750,
+        "main-es2015": 432199,
```
Comment (gkalpak) [inline goldens/size-tracking/aio-payloads.json:24]: The actual increase here is 277B (from [431922][1] to [432199][2]).

[1]: https://app.circleci.com/pipelines/github/angular/angular/24192/workflows/1f86d952-a813-4e00-8aba-11c2420c18dc/jobs/850934
[2]: https://app.circleci.com/pipelines/github/angular/angular/24189/workflows/03d37062-e9d7-401e-a563-7a386db54994/jobs/850878

**Comment by atscott** [comment]
[global presubmit](https://test.corp.google.com/ui#id=OCL:340874556:BASE:340874705:1604599375421:32d7e440)

**Pull request closed without merging**